jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -103,57 +103,83 @@ ecma_builtin_string_prototype_object_value_of (ecma_value_t this_arg) /**< this
 } /* ecma_builtin_string_prototype_object_value_of */
 
 /**
- * The String.prototype object's 'charAt' routine
- *
- * See also:
- *          ECMA-262 v5, 15.5.4.4
+ * Helper function for the String.prototype object's 'charAt' and charCodeAt' routine
  *
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 static ecma_value_t
-ecma_builtin_string_prototype_object_char_at (ecma_value_t this_arg, /**< this argument */
-                                              ecma_value_t arg) /**< routine's argument */
+ecma_builtin_string_prototype_char_at_helper (ecma_value_t this_arg, /**< this argument */
+                                              ecma_value_t arg, /**< routine's argument */
+                                              bool charcode_mode) /**< routine mode */
 {
-  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
-
   /* 1 */
-  ECMA_TRY_CATCH (check_coercible_val,
-                  ecma_op_check_object_coercible (this_arg),
-                  ret_value);
-
-  /* 2 */
-  ECMA_TRY_CATCH (to_string_val,
-                  ecma_op_to_string (this_arg),
-                  ret_value);
+  ecma_value_t check_coercible_val = ecma_op_check_object_coercible (this_arg);
+  if (ECMA_IS_VALUE_ERROR (check_coercible_val))
+  {
+    return check_coercible_val;
+  }
+  ecma_free_value (check_coercible_val);
 
   /* 3 */
-  ECMA_OP_TO_NUMBER_TRY_CATCH (index_num,
-                               arg,
-                               ret_value);
+  ecma_number_t index_num;
+  ecma_value_t to_num_result = ecma_get_number (arg, &index_num);
+  if (unlikely (!ecma_is_value_empty (to_num_result)))
+  {
+    return to_num_result;
+  }
+  ecma_free_value (to_num_result);
+
+  /* 2 */
+  ecma_value_t to_string_val = ecma_op_to_string (this_arg);
+  if (ECMA_IS_VALUE_ERROR (to_string_val))
+  {
+    return to_string_val;
+  }
 
   /* 4 */
   ecma_string_t *original_string_p = ecma_get_string_from_value (to_string_val);
   const ecma_length_t len = ecma_string_get_length (original_string_p);
 
   /* 5 */
-  if (index_num < 0 || index_num >= len || !len)
-  {
-    ret_value = ecma_make_magic_string_value (LIT_MAGIC_STRING__EMPTY);
-  }
-  else
+  // When index_num is NaN, then the first two comparisons are false
+  if (index_num < 0 || index_num >= len || (ecma_number_is_nan (index_num) && !len))
   {
-    /* 6 */
-    ecma_char_t new_ecma_char = ecma_string_get_char_at_pos (original_string_p, ecma_number_to_uint32 (index_num));
-    ret_value = ecma_make_string_value (ecma_new_ecma_string_from_code_unit (new_ecma_char));
+    ecma_free_value (to_string_val);
+    return charcode_mode ? ecma_make_nan_value ()
```

Comment:
Use parenthesis, so vera++ rules does not complain about wrong indentation.
```
return (charcode_mode ? ecma_make_nan_value ()
                      : ecma_make_magic_string_value (LIT_MAGIC_STRING__EMPTY));
```

Reply:
Thanks, that worked.